Added mean_absolute_percentage_error in metrics fixes #10708

ashutosh1919 opened the pull request:
I have faced problem when working with any quantitative modelling algorithm. Scikit-learn doesn't have any function such **mean_absolute_percentage_error**. Everytime, when I want to use this function, then I have to implement it myself. But, if it is included in sklearn itself then it will be great. I have written code for this function and I have also added tests for it. @agramfort , Please Review my PR.
![mape](https://user-images.githubusercontent.com/20843596/65117948-26611180-da08-11e9-837d-68af24949e57.png)
This PR Fixes #10708 .  This PR is continuation of the work done by @mohamed-ali in #10711 PR.

ashutosh1919 committed:
Added mean_absolute_percentage_error in metrics

ashutosh1919 committed:
Added mean_absolute_percentage_error in metrics

ashutosh1919 committed:
Added mean_absolute_percentage_error in metrics

agramfort commented on sklearn/metrics/regression.py:
```diff
@@ -189,6 +189,78 @@ def mean_absolute_error(y_true, y_pred,
+    y_type, y_true, y_pred, multioutput = _check_reg_targets(
+        y_true, y_pred, multioutput)
+    check_consistent_length(y_true, y_pred, sample_weight)
+    mape = np.abs((y_pred - y_true) / (1 + np.abs(y_true)))
```
why do you put a 1 at the denominator ? cf https://en.wikipedia.org/wiki/Mean_absolute_percentage_error

ashutosh1919 replied:
It's check. If I don't put 1 in denomenator, then it will return **inf** when y_true is **0** and then the value returned by mean_absolute_percentage_error will be **nan**. To avoid this, I am using 1 which won't abruptly deviate the error value and also will be consistent. @agramfort

agramfort commented:
> It's check. If I don't put 1 in denomenator, then it will return inf when
y_true is 0 and then the value
> returned by mean_absolute_percentage_error will be nan. To avoid this, I
am using 1 which won't
> abruptly deviate the error value and also will be consistent. @agramfort

I would not do this. To me it's a feature not a bug that you can NaN / Inf
if you have any y that is 0.
If this happens in practice I presume it's bug in the dataset.

ashutosh1919 commented:
> It's check. If I don't put 1 in denomenator, then it will return inf when
> y_true is 0 and then the value
> returned by mean_absolute_percentage_error will be nan. To avoid this, I
> am using 1 which won't
> abruptly deviate the error value and also will be consistent. @agramfort
> I would not do this. To me it's a feature not a bug that you can NaN / Inf if you have any y that is 0. If this happens in practice I presume it's bug in the dataset.

I was not using 1 previously but when I wrote test script for the function, one of the test case was given like below : 
```
y_true = np.array([[1, 0, 0, 1], [0, 1, 1, 1], [1, 1, 0, 1]])
y_pred = np.array([[0, 0, 0, 1], [1, 0, 1, 1], [0, 0, 0, 1]])
```
Above input is completely valid input and looks like the output of the classifier. If I don't put 1 in denomenator, it will fail and gives me **nan**. But if I put 1 then it gives me correct answer **8.99** which I needed at last.

jnothman commented:
See an alternative contribution of this feature in #10711. I've not yet reviewed where it got stuck

ashutosh1919 commented:
> See an alternative contribution of this feature in #10711. I've not yet reviewed where it got stuck

I have seen the pending PR, **should I raise error when any y_true is 0?** If that is the case then I will write seperate test function for evaluating **mean_absolute_percentage_error** because the tests written for other error functions won't work for this one. Should I do that? or should I search for some better approach? Awaiting for reply @agramfort @jnothman

ashutosh1919 commented:
> > See an alternative contribution of this feature in #10711. I've not yet reviewed where it got stuck
> 
> I have seen the pending PR, **should I raise error when any y_true is 0?** If that is the case then I will write seperate test function for evaluating **mean_absolute_percentage_error** because the tests written for other error functions won't work for this one. Should I do that? or should I search for some better approach? Awaiting for reply @agramfort @jnothman

ashutosh1919 commented:
@jnothman @agramfort , Please look at **this one**. This is implementation of mean_absolute_percentage_error loss in tensorflow. **They have clipped the smallest value to EPSILON.** I think this is more robust approach than giving error. This will also work with every test case. If we raise on finding zero, most of the standard binary classification tasks will give error. But if we clip denominator to **EPSILON**, then it will work in every case.

ashutosh1919 committed:
Added MAPE

ashutosh1919 committed:
Added MAPE

ashutosh1919 committed:
Added MAPE

ashutosh1919 committed:
Added MAPE

ashutosh1919 committed:
Added MAPE

ashutosh1919 committed:
MAPE implementation changed

ashutosh1919 committed:
MAPE implementation changed

ashutosh1919 commented:
@amueller, @mohamed-ali, @jnothman, @agramfort 
I have implemented MAPE using following formula and considering the comments which you have specified in Issue #10708 and PR #10711. 
![mape](https://user-images.githubusercontent.com/20843596/65146950-b28c2c80-da3a-11e9-8f39-ebd4cfc55dc0.png)
Where **epsilon is the smallest number greater than zero**.
Implementation is based on **np.finfo(float).eps** which has value of **2.220446049250313e-16**.
Also I have found implementation of **MAPE in tensorflow** losses [here](https://github.com/tensorflow/tensorflow/blob/ca123bc4199a42c65dcc817b6c771fa613171e08/tensorflow/python/keras/losses.py#L792) as shown below : 
```
def mean_absolute_percentage_error(y_true, y_pred):  # pylint: disable=missing-docstring
  y_pred = ops.convert_to_tensor(y_pred)
  y_true = math_ops.cast(y_true, y_pred.dtype)
  diff = math_ops.abs(
      (y_true - y_pred) / K.clip(math_ops.abs(y_true), K.epsilon(), None))
  return 100. * K.mean(diff, axis=-1)
```
Above code shows that they have **clipped the denominator value to epsilon lower bound**. 
I have implemented the same thing using numpy.

**Advantages of this logic :** 
(i) Code will not give unexpected **inf** or **nan** value when used anywhere instead it will give very high error so that the developer will get to know that somewhere in the denominator, zero is encountered. It won't crash the model at any time instead it will show high error.
(ii) This logic doesn't require changing the test functions for this particular metrics function. Instead, we can test this function using the existing test cases and no other test cases are needed.
(iii) It doesn't violate the definition of MAPE which is shown in the Wikipedia and it won't deviate the error in the significant decimal points.

Current Logic Passes all test cases and implementation is robust as per tensorflow.

agramfort commented on sklearn/metrics/regression.py:
```diff
@@ -189,6 +189,80 @@ def mean_absolute_error(y_true, y_pred,
+    a_max = np.max(np.abs(y_true))
+    a_max = max(epsilon, a_max)
+    denom = np.clip(np.abs(y_true), epsilon, a_max)
+    mape = np.abs((y_pred - y_true) / denom)
```
```suggestion
    mape = np.abs((y_pred - y_true) / np.maximum(np.abs(y_true), epsilon)
```

no need for a_max and clip

ashutosh1919 replied:
Thank you for your suggestion @agramfort. It is really a better way to do it. I have made the changes and removed unnecessary code. Please review PR.

ashutosh1919 committed:
Removed Clip and applied np.maximum

agramfort commented:
@ashutosh1919 you need to update classes.rst and doc/modules/model_evaluation.rst and scorer.py

a good way to find this out is to do a git grep on the related function eg `git grep median_absolute_error`

ashutosh1919 committed:
MAPE Added in Docs

ashutosh1919 commented:
> @ashutosh1919 you need to update classes.rst and doc/modules/model_evaluation.rst and scorer.py
> 
> a good way to find this out is to do a git grep on the related function eg `git grep median_absolute_error`

I have added the doc for MAPE as per your instructions. @agramfort, Please take a look at doc. It is passes all tests.

mohamed-ali commented:
@agramfort, @amueller @jnothman 

I don't know how contribution are orchestrated in sklearn, but the issue https://github.com/scikit-learn/scikit-learn/issues/10708 has been addressed with my PR https://github.com/scikit-learn/scikit-learn/pull/10711. 

The PR was approved by @jnothman, then remained unmerged for more than one year, and I have been recently asked by @amueller to synchronized with master and now I see this PR being created to address the exact same issue? 

@agramfort @ashutosh1919 Could you please close the PR I created, or at least let me know there beforehand so I don't waste more time.

Thanks.

ashutosh1919 commented:
> @agramfort, @amueller @jnothman
> 
> I don't know how contribution are orchestrated in sklearn, but the issue #10708 has been addressed with my PR #10711.
> 
> The PR was approved by @jnothman, then remained unmerged for more than one year, and I have been recently asked by @amueller to synchronize it with master, but now I see this PR being created to address the exact same issue?
> 
> @agramfort @ashutosh1919 Could you please close the PR I created, or at least let me know there beforehand so I don't waste more time.
> 
> Thanks.

Sorry @mohamed-ali, At the time when I raised my PR, I didn't know that issue for the same is raised before and PR is already there on it.  Thanks to @jnothman that he informed about your PR and also about the issue. By seeing the last update on your PR, I thought that there has not been any update since then. I am extremely sorry for that. But I have mentioned you in [here](https://github.com/scikit-learn/scikit-learn/pull/15007#issuecomment-532659830) so that you stay aware about the same issue. Hands over to @agramfort, Please close one of the PR (Mine or his) on same issue.

mohamed-ali commented:
@ashutosh1919 You can take it over, I will focus on other things. Just make sure to mention my old PR so people know you are continuing the work.

agramfort commented on doc/modules/model_evaluation.rst:
```diff
@@ -95,6 +95,7 @@ Scoring                           Function
 'neg_root_mean_squared_error'     :func:`metrics.mean_squared_error`
 'neg_mean_squared_log_error'      :func:`metrics.mean_squared_log_error`
 'neg_median_absolute_error'       :func:`metrics.median_absolute_error`
+'neg_mean_absolute_percentage_error' :func:`metrics.mean_absolute_percentage_error`
```
this will break rendering of the rst file.

ashutosh1919 replied:
removed.

agramfort commented on doc/modules/model_evaluation.rst:
```diff
@@ -1936,6 +1937,37 @@ function::
+it is directly understood by the users who don't have technical background.
+It is modification of `mean_absolute_error`. The loss is calculated by taking
+average of all differences between the target and the prediction divided by
+the target and multiplying resulting term by 100.0 to convert it into percentage.
```
explain what it does : "penalize errors in percentage" vs absolute value for MAE. It is therefore blind to global scaling of the target as long as the prediction is scaled like the true y. Try to rephrase or just copy paste
from some wikipedia page or similar.

ashutosh1919 replied:
changed

agramfort commented on doc/whats_new/v0.16.rst:
```diff
@@ -118,6 +118,9 @@ New features
   By `Gael Varoquaux`_ and :user:`Florian Wilhelm <FlorianWilhelm>`.
 
+- Added :func:`metrics.mean_absolute_percentage_error`, a robust metric.
+  By :user:`Ashutosh Hathidara <ashutosh1919>` and `Alexandre Gramfort`_.
```
don't add me here. I just review

ashutosh1919 replied:
changed.

ashutosh1919 committed:
Changed model_evaluation descriptions and other changes

ashutosh1919 committed:
Resolving error

agramfort commented on doc/modules/model_evaluation.rst:
```diff
@@ -95,7 +95,6 @@ Scoring                           Function
 'neg_root_mean_squared_error'     :func:`metrics.mean_squared_error`
 'neg_mean_squared_log_error'      :func:`metrics.mean_squared_log_error`
 'neg_median_absolute_error'       :func:`metrics.median_absolute_error`
-'neg_mean_absolute_percentage_error' :func:`metrics.mean_absolute_percentage_error`
```
don't remove it just reformat the table so it renders fine.

ashutosh1919 replied:
changed as per my understanding. aligned with other lines.

ashutosh1919 committed:
model_evaluation table changed

ashutosh1919 committed:
model_evaluation table changed

ashutosh1919 committed:
model_evaluation table changed

ashutosh1919 commented:
@agramfort , I have only added one line in model_evaluation.rst and that is the doc part. I don't know why this is failing in C++ analysis. Can you please suggest me changes to solve C++ analysis build fail error?

ashutosh1919 committed:
Merge remote-tracking branch 'upstream/master'

ashutosh1919 committed:
Merge branch 'master' of https://github.com/ashutosh1919/scikit-learn

ashutosh1919 commented:
Synchronised master  of forked repo with master of parent repo. This solves the problem. @agramfort, Please review this PR now.

mohamed-ali commented on sklearn/metrics/regression.py:
```diff
@@ -189,6 +189,77 @@ def mean_absolute_error(y_true, y_pred,
+    mape = np.abs(y_pred - y_true) / np.maximum(np.abs(y_true), epsilon)
+    output_errors = np.average(mape,
+                               weights=sample_weight, axis=0) * 100.0
+    if isinstance(multioutput, str):
```
The following is equivalent with less code:

```
    if isinstance(multioutput, str):
        if multioutput == 'raw_values':
            return output_errors
    return np.average(output_errors, weights=None)
```

ashutosh1919 replied:
implemented as considering the other error function into consideration and syntax is consistent with other error functions. Time complexity is pretty much same in your code and in mine. You are making two if, I am making if-elif.

ashutosh1919 committed:
Changed Doc line

ashutosh1919 commented on sklearn/metrics/regression.py:
```diff
@@ -189,6 +189,77 @@ def mean_absolute_error(y_true, y_pred,
+
+        'uniform_average' :
+            Errors of all outputs are averaged with uniform weight.
+
```
checked. There is no pep8 error and there are already two lines skipped in other error function at the same location in same file.

ashutosh1919 commented on doc/modules/model_evaluation.rst:
```diff
@@ -1936,6 +1938,37 @@ function::
+The :func: The `mean_absolute_percentage_error` (MAPE), also known as mean absolute
+percentage deviation (MAPD), is a measure of prediction accuracy of a forecasting
+method in statistics, for example in trend estimation, also used as a loss function
+for regression problems in machine learning. It is therefore blind to global scaling
```
Changed.

ashutosh1919 committed:
Merge remote-tracking branch 'upstream/master'

agramfort commented:
@ashutosh1919 you need to fix conflicts with rebase or merging master

ashutosh1919 committed:
Sync with remote

ashutosh1919 committed:
metrics init file changed

ashutosh1919 committed:
test_regression resolved

ashutosh1919 committed:
test_regression resolved

ashutosh1919 commented:
> @ashutosh1919 you need to fix conflicts with rebase or merging master

@agramfort , Please check the changes, there are no conflicts now. Please add it.

agramfort commented on doc/modules/model_evaluation.rst:
```diff
@@ -91,6 +91,8 @@ Scoring                           Function
 'max_error'                       :func:`metrics.max_error`
 'neg_mean_absolute_error'         :func:`metrics.mean_absolute_error`
+'neg_mean_absolute_percentage_error'
+                                  :func:`metrics.mean_absolute_percentage_error`
```
this will break rendered doc

ashutosh1919 replied:
Please suggest the change. I have tried several ways but every time you have given this comment. Please suggest code snippet.

agramfort commented on doc/modules/model_evaluation.rst:
```diff
@@ -1955,6 +1957,37 @@ function::
+------------------------------
+The :func: The `mean_absolute_percentage_error` (MAPE), also known as mean absolute
+percentage deviation (MAPD), is a measure of prediction accuracy of a forecasting
+method in statistics, for example in trend estimation, also used as a loss function
```
remove refs to forecasting. scikit-learn is not about time series forecasting although it can used with some overheads. Just say it's for regression

ashutosh1919 replied:
DONE

ashutosh1919 committed:
Resolving Forecasting text

ashutosh1919 committed:
Render error

jnothman commented:
Please fix the ci failures, @ashutosh1919

ashutosh1919 committed:
Merge remote-tracking branch 'upstream/master'

ashutosh1919 committed:
Resolving render errors

ashutosh1919 committed:
Render doc error

ashutosh1919 committed:
Resolving render error

ashutosh1919 committed:
Resolving render doc

ashutosh1919 committed:
Resolving render doc

ashutosh1919 commented:
> Please fix the ci failures, @ashutosh1919

@jnothman , Please suggest me code for rendering failure. I have tried many ways and all failed. It is showing Malformed table. Please help.

ashutosh1919 committed:
Resolving render doc

ashutosh1919 commented:
@agramfort or @jnothman, Please review and merge this PR as soon as possible. If there are more commits added then I again need to rebase.

agramfort commented on doc/modules/model_evaluation.rst:
```diff
@@ -1959,6 +1959,38 @@ function::
+Where :math:`\epsilon` is a very small number greater than zero and kept in order to avoid
+the crash when y is zero.
+
+The :func:`mean_absolute_percentage_error` support multioutput.
```
```suggestion
The :func:`mean_absolute_percentage_error` function supports multioutput.
```

agramfort commented on doc/whats_new/v0.16.rst:
```diff
@@ -118,6 +118,9 @@ New features
 - Added :func:`metrics.median_absolute_error`, a robust metric.
   By `Gael Varoquaux`_ and :user:`Florian Wilhelm <FlorianWilhelm>`.
 
+- Added :func:`metrics.mean_absolute_percentage_error`, a robust metric.
```
robust metric has been copy pasted while it's not relevant here

agramfort commented on sklearn/metrics/_regression.py:
```diff
@@ -189,6 +190,77 @@ def mean_absolute_error(y_true, y_pred,
+
+    Parameters
+    ----------
+    y_true : array-like of shape = (n_samples) or (n_samples, n_outputs)
```
```suggestion
    y_true : array-like of shape = (n_samples,) or (n_samples, n_outputs)
```

I would write is this way. Not sure it's consistent with what is done elsewhere

agramfort commented on sklearn/metrics/_regression.py:
```diff
@@ -189,6 +190,77 @@ def mean_absolute_error(y_true, y_pred,
+    y_true : array-like of shape = (n_samples) or (n_samples, n_outputs)
+        Ground truth (correct) target values.
+
+    y_pred : array-like of shape = (n_samples) or (n_samples, n_outputs)
```
```suggestion
    y_pred : array-like of shape = (n_samples,) or (n_samples, n_outputs)
```

agramfort commented on sklearn/metrics/_regression.py:
```diff
@@ -189,6 +190,77 @@ def mean_absolute_error(y_true, y_pred,
+    y_pred : array-like of shape = (n_samples) or (n_samples, n_outputs)
+        Estimated target values.
+
+    sample_weight : array-like of shape = (n_samples), optional
```
```suggestion
    sample_weight : array-like of shape = (n_samples,), optional
```

agramfort commented on sklearn/metrics/_regression.py:
```diff
@@ -189,6 +190,77 @@ def mean_absolute_error(y_true, y_pred,
+        Sample weights.
+
+    multioutput : string in ['raw_values', 'uniform_average']
+        or array-like of shape (n_outputs)
```
```suggestion
        or array-like of shape (n_outputs,)
```

does it render well with sphinx? I suspect this line is appear in the description text

ashutosh1919 replied:
Yeah it renders perfectly with sphinx. I have checked it out.

agramfort commented on sklearn/metrics/_regression.py:
```diff
@@ -189,6 +190,77 @@ def mean_absolute_error(y_true, y_pred,
+    >>> y_pred = [[0, 2], [-1, 2], [8, -5]]
+    >>> mean_absolute_percentage_error(y_true, y_pred)
+    55.15...
+    >>> mean_absolute_percentage_error(y_true, y_pred, multioutput=[0.3, 0.7])
```
do says that multioutput should be a string. It's here a list.

ashutosh1919 replied:
multioutput can take string or list as input.

Read the description given above the implementation function : 

multioutput : string in ['raw_values', 'uniform_average']
        or array-like of shape (n_outputs)
        Defines aggregating of multiple output values.
        Array-like value defines weights used to average errors.

        'raw_values' :
            Returns a full set of errors in case of multioutput input.

        'uniform_average' :
            Errors of all outputs are averaged with uniform weight.

So, it can use string from ['raw_values', 'uniform_average'] or it can use list

ashutosh1919 committed:
Update doc/modules/model_evaluation.rst

Co-Authored-By: Alexandre Gramfort <alexandre.gramfort@m4x.org>

ashutosh1919 committed:
Update sklearn/metrics/_regression.py

Co-Authored-By: Alexandre Gramfort <alexandre.gramfort@m4x.org>

ashutosh1919 committed:
Update sklearn/metrics/_regression.py

Co-Authored-By: Alexandre Gramfort <alexandre.gramfort@m4x.org>

ashutosh1919 committed:
Applying suggested changes

ashutosh1919 committed:
Merge remote-tracking branch 'upstream/master'

ashutosh1919 committed:
Applying suggested changes

ashutosh1919 committed:
Applying suggested changes

ashutosh1919 committed:
Applying suggested changes

ashutosh1919 committed:
Applying suggested changes

ashutosh1919 committed:
Applying suggested changes

ashutosh1919 commented:
@agramfort, I have changed the code as per your suggestions. Just take a look at it.

agramfort commented on doc/modules/model_evaluation.rst:
```diff
@@ -1959,6 +1959,38 @@ function::
+
+  \text{MAPE}(y, \hat{y}) = \frac{\mid{y} - \hat{y}\mid}{\text{MAX}(\epsilon, \mid{y}\mid)}
+
+Where :math:`\epsilon` is a very small number greater than zero and kept in order to avoid
```
```suggestion
where :math:`\epsilon` is a very small number greater than zero to avoid
```

agramfort commented on doc/modules/model_evaluation.rst:
```diff
@@ -1959,6 +1959,38 @@ function::
+
+.. math::
+
+  \text{MAPE}(y, \hat{y}) = \frac{\mid{y} - \hat{y}\mid}{\text{MAX}(\epsilon, \mid{y}\mid)}
```
```suggestion
  \text{MAPE}(y, \hat{y}) = \frac{\mid y - \hat{y} \mid}{\text{MAX}(\epsilon, \mid y \mid)}
```

ashutosh1919 committed:
Merge remote-tracking branch 'upstream/master'

ashutosh1919 committed:
Update doc/modules/model_evaluation.rst

Co-Authored-By: Alexandre Gramfort <alexandre.gramfort@m4x.org>

ashutosh1919 committed:
Update doc/modules/model_evaluation.rst

Co-Authored-By: Alexandre Gramfort <alexandre.gramfort@m4x.org>

ashutosh1919 committed:
Update doc/modules/model_evaluation.rst

Co-Authored-By: Alexandre Gramfort <alexandre.gramfort@m4x.org>

ashutosh1919 committed:
Update doc/modules/model_evaluation.rst

Co-Authored-By: Alexandre Gramfort <alexandre.gramfort@m4x.org>

ashutosh1919 committed:
Update doc/modules/model_evaluation.rst

Co-Authored-By: Alexandre Gramfort <alexandre.gramfort@m4x.org>

ashutosh1919 committed:
Update doc/modules/model_evaluation.rst

Co-Authored-By: Alexandre Gramfort <alexandre.gramfort@m4x.org>

ashutosh1919 committed:
Update doc/modules/model_evaluation.rst

Co-Authored-By: Alexandre Gramfort <alexandre.gramfort@m4x.org>

ashutosh1919 committed:
Doc Too long line error resolved

ashutosh1919 commented:
@agramfort , Please review PR.

ogrisel commented on sklearn/metrics/_regression.py:
```diff
@@ -189,6 +190,77 @@ def mean_absolute_error(y_true, y_pred,
+    y_type, y_true, y_pred, multioutput = _check_reg_targets(
+        y_true, y_pred, multioutput)
+    check_consistent_length(y_true, y_pred, sample_weight)
+    epsilon = np.finfo(float).eps
```
Please put the precision explicitly:  `np.finfo(np.float64).eps`.

Actually I wonder if we should make it depend on the dtype of `y_true`:

```python
epsilon = np.finfo(y_true.dtype).eps
```

Any opinion? Maybe always using `np.float64` is less likely to cause hard to understand, unexpected behaviors though.

ashutosh1919 replied:
I agree with your opinion. It's good to have dtype of y_true than taking float64 for every case.

ashutosh1919 replied:
I think np.float64 makes more sense to me because in most of other datatype cases, the datatype would be integer, and it will fail finding the eps for that since eps has to be float real number. @ogrisel, Your comments will be highly appreciated on this.

ogrisel replied:
Let keep `np.finfo(np.float64).eps` then.

ashutosh1919 replied:
@ogrisel  ,I have kept np.float64 in current commit. Please review the complete code then.

ogrisel commented on sklearn/metrics/tests/test_regression.py:
```diff
@@ -32,6 +33,10 @@ def test_regression_metrics(n_samples=50):
+    assert_almost_equal(np.around(mean_absolute_percentage_error(y_true,
+                                                                 y_pred),
+                                  decimals=2),
+                        9007199254741000.0, decimal=-13)
```
This is a weird test. `mean_absolute_percentage_error` is really not mean to be evaluated on problem where `y_true` can have exactly zero values. I would rather not make any assertion on the exact value you get for this pathological case because nobody would care about the actual value as long as it's a finite yet large positive number.

Maybe just do something like:

```python
# Because there is a 0 in `y_true`, it does not really make sense to use MAPE for
# this regression task. Here we just check that the score is a large positive yet
# finite value:
mape = mean_absolute_percentage_error(y_true, y_pred)
assert np.isfinite(mape)
assert mape > 1e6  # arbitrary large enough positive value

# If the two values are equal and strictly positive, the expected MAPE is 0.
mape = mean_absolute_percentage_error(y_true + 1, y_pred)
assert  mape == pytest.approx(0)
```

ashutosh1919 replied:
done

ogrisel commented on sklearn/metrics/tests/test_regression.py:
```diff
@@ -86,6 +91,9 @@ def test_multioutput_regression():
 
+    error = np.around(mean_absolute_percentage_error(y_true, y_pred),
+                      decimals=2)
+    assert_almost_equal(error, 375e+14, decimal=-14)
```
Same remark as above for this assertion.

ashutosh1919 replied:
done

ogrisel reviewed (commented):
Please add a test dedicated to highlight the expected behavior of MAPE:

```python
y_true = [1., 10., 100.]
y_pred = [0.8, 1.2, 80.]
mape = mean_absolute_percentage_error(y_true, y_pred)
assert mape == pytest.approx(20)
```

Maybe this case could actually be adapted as a doctest in the docstring of the `mean_absolute_percentage_error` function itself.

Please also add a scorer object named `"neg_mean_absolute_percentage_error"` that returns the negative MAPE. Have a look at how the `"neg_mean_squared_error"` scorer is defined in `sklearn/metrics/_scorer.py` as an example.

ogrisel commented on doc/modules/model_evaluation.rst:
```diff
@@ -1959,6 +1959,37 @@ function::
+function::
+
+  >>> from sklearn.metrics import mean_absolute_percentage_error
+  >>> y_true = [3, -0.5, 2, 7]
```
I would rather not put negative value in the canonical doctest for this scoring metric. While MAPE is technically defined in this case, MAPE is really not intended to be used with smalle `y_true` values that oscillate around 0.

I would rather highlight is behavior with `y_true` values that take widely scaled positive values, e.g. `[1., 10, 1e6]` and that approximately follows those scales `y_pred = [0.9, 15, 1.2e6]`.

ashutosh1919 replied:
done

ogrisel reviewed (commented):
Also I think we should also add `mean_absolute_relative_error` that computes the same as `MAPE` but without multiplying the score by 100.

WDYT?

ogrisel commented on doc/whats_new/v0.16.rst:
```diff
@@ -118,6 +118,9 @@ New features
   By `Gael Varoquaux`_ and :user:`Florian Wilhelm <FlorianWilhelm>`.
 
+- Added :func:`metrics.mean_absolute_percentage_error`.
+  By :user:`Ashutosh Hathidara <ashutosh1919>`.
```
The next release is going 0.23. So the changelog entry should be added in v0.23.rst instead.

ashutosh1919 replied:
done

ogrisel commented:
Sorry @mohamed-ali I had not see your original PR #10711 either when I started to review this PR.

Contrary to what is state in the description of this PR, this is not continuation of #10711. In particular #10711 had the following things that are missing in this PR:

- the scorer object;
- tests that makes more sense.

However this PR has nice additions:

- multioutput support;
- epsilon in the denominator for convenience.

@ashutosh1919 it would have been better to start from the commits of @mohamed-ali's branch to then add your own contributions of top of it but I guess now is too late to do so.

Could you please add the missing features from @mohamed-ali's PR and update the whats_new entry to give proper credit to @mohamed-ali's contribution?

ogrisel commented:
> Also I think we should also add mean_absolute_relative_error that computes the same as MAPE but without multiplying the score by 100.

@jnothman suggested something similar in:

https://github.com/scikit-learn/scikit-learn/issues/10708#issuecomment-368674055

ashutosh1919 commented:
@ogrisel , Sorry for being late. I have looked at your reviews on this PR. It looks good to me. I will change it as per your comments. Also, for `mean_absolute_relative_error`, i will raise another PR soon. Please give me some time. Will do everything soon.

ashutosh1919 committed:
Merge remote-tracking branch 'upstream/master'

ashutosh1919 committed:
datatype changed and made compatible to y_true

ashutosh1919 committed:
Added scorer

ashutosh1919 committed:
eps datatype changed to np.float64

ashutosh1919 committed:
test_regression.py is changed to more meaningful test cases

ashutosh1919 committed:
test_regression.py is changed to more meaningful test cases

ashutosh1919 committed:
Resolving errors related to scorer tests in test_common.py and _scorer.py

ashutosh1919 committed:
Updated v0.23.rst in whats_new

ashutosh1919 committed:
resolving errors of mape scorer

ashutosh1919 committed:
resolving errors of mape scorer

ashutosh1919 committed:
resolving errors of mape scorer

ashutosh1919 committed:
resolving errors of mape scorer

ashutosh1919 committed:
modified test case in model_evaluation.rst

ashutosh1919 commented:
@ogrisel , Please review PR now. I have made changes as you said.

ogrisel reviewed (commented):
Here is another batch of comments.

@jnothman what do you think of the suggestion to introduce `mean_absolute_relative_error` that does not multiply by 100.?

ashutosh1919 committed:
modified doc and code as per second batch comments

ashutosh1919 committed:
Resolving r2_scorer object error

ashutosh1919 committed:
Resolving r2_scorer object error

ashutosh1919 committed:
Merge remote-tracking branch 'upstream/master'

ashutosh1919 committed:
Resolving r2_scorer object error

ashutosh1919 committed:
Resolving r2_scorer object error

ashutosh1919 commented:
@ogrisel , Changed the files according to 2nd batch of comments. Please take a look.

ashutosh1919 commented on doc/modules/model_evaluation.rst:
```diff
@@ -1958,6 +1958,37 @@ function::
+  \text{MAPE}(y, \hat{y}) = \frac{\mid y - \hat{y} \mid}{\text{MAX}(\epsilon, \mid y \mid)}
+
+where :math:`\epsilon` is a very small number greater than zero to avoid
+the crash when y is zero.
```
done

ashutosh1919 commented on doc/modules/model_evaluation.rst:
```diff
@@ -1958,6 +1958,37 @@ function::
+  >>> y_pred = [0.9, 15, 1.2e6]
+  >>> mean_absolute_percentage_error(y_true, y_pred)
+  26.66...
+
```
done

ashutosh1919 commented on sklearn/metrics/_regression.py:
```diff
@@ -189,6 +191,76 @@ def mean_absolute_error(y_true, y_pred,
+        If multioutput is 'uniform_average' or an ndarray of weights, then the
+        weighted average of all output errors is returned.
+
+        MAPE output is non-negative floating point. The best value is 0.0.
```
done

ashutosh1919 commented on sklearn/metrics/_scorer.py:
```diff
@@ -687,6 +689,7 @@ def make_scorer(score_func, greater_is_better=True, needs_proba=False,
                max_error=max_error_scorer,
                neg_median_absolute_error=neg_median_absolute_error_scorer,
                neg_mean_absolute_error=neg_mean_absolute_error_scorer,
+               neg_mean_absolute_percentage_error=neg_mape_scorer,
```
done

ashutosh1919 commented on sklearn/metrics/tests/test_common.py:
```diff
@@ -1371,7 +1373,11 @@ def test_thresholded_multilabel_multioutput_permutations_invariance(name):
-        assert_almost_equal(score, current_score)
+        if metric == mean_absolute_percentage_error:
+            assert np.isfinite(current_score)
+            assert current_score > 1e6
```
done

ashutosh1919 commented on sklearn/metrics/tests/test_regression.py:
```diff
@@ -254,12 +266,15 @@ def test_regression_custom_weights():
+    assert_almost_equal(mapew, 16.68, decimal=2)
     assert_almost_equal(rw, 0.94, decimal=2)
     assert_almost_equal(evsw, 0.94, decimal=2)
 
```
done

ashutosh1919 commented on doc/whats_new/v0.23.rst:
```diff
@@ -217,6 +217,12 @@ Changelog
+  the associated scorer for regression problems. :issue:`10708` fixed with the
+  PR :pr:`15007` by :user:`Ashutosh Hathidara <ashutosh1919>`. The scorer and
+  some practical test cases were taken from PR :pr:`10711` by
+  :user:`Mohamed Ali Jamaoui <mohamed-ali>`.
```
done

ashutosh1919 commented on sklearn/metrics/_scorer.py:
```diff
@@ -623,6 +623,8 @@ def make_scorer(score_func, greater_is_better=True, needs_proba=False,
 neg_mean_absolute_error_scorer = make_scorer(mean_absolute_error,
                                              greater_is_better=False)
+neg_mape_scorer = make_scorer(mean_absolute_percentage_error,
+                              greater_is_better=False)
```
done

ashutosh1919 committed:
Conflict Resolved

ashutosh1919 committed:
Added changes of conflict

thomasjpfan commented on sklearn/metrics/_regression.py:
```diff
@@ -189,6 +191,78 @@ def mean_absolute_error(y_true, y_pred,
+
+    Parameters
+    ----------
+    y_true : array-like of shape = (n_samples,) or (n_samples, n_outputs)
```
```suggestion
    y_true : array-like of shape (n_samples,) or (n_samples, n_outputs)
```

thomasjpfan commented on sklearn/metrics/_regression.py:
```diff
@@ -189,6 +191,78 @@ def mean_absolute_error(y_true, y_pred,
+    y_true : array-like of shape = (n_samples,) or (n_samples, n_outputs)
+        Ground truth (correct) target values.
+
+    y_pred : array-like of shape = (n_samples,) or (n_samples, n_outputs)
```
```suggestion
    y_pred : array-like of shape (n_samples,) or (n_samples, n_outputs)
```

thomasjpfan commented on sklearn/metrics/_regression.py:
```diff
@@ -189,6 +191,78 @@ def mean_absolute_error(y_true, y_pred,
+    y_pred : array-like of shape = (n_samples,) or (n_samples, n_outputs)
+        Estimated target values.
+
+    sample_weight : array-like of shape = (n_samples,), optional
```
```suggestion
    sample_weight : array-like of shape (n_samples,), default=None
```

thomasjpfan commented on sklearn/metrics/_regression.py:
```diff
@@ -189,6 +191,78 @@ def mean_absolute_error(y_true, y_pred,
+    sample_weight : array-like of shape = (n_samples,), optional
+        Sample weights.
+
+    multioutput : string in ['raw_values', 'uniform_average'] or list.
```
```suggestion
    multioutput : {'raw_values', 'uniform_average'} or array-like
```

thomasjpfan commented on sklearn/metrics/_scorer.py:
```diff
@@ -682,34 +685,36 @@ def make_scorer(score_func, greater_is_better=True, needs_proba=False,
-               adjusted_mutual_info_score=adjusted_mutual_info_scorer,
-               normalized_mutual_info_score=normalized_mutual_info_scorer,
-               fowlkes_mallows_score=fowlkes_mallows_scorer)
+SCORERS = dict(
```
Can we minimize the diff by only adding `neg_mean_absolute_percentage_error_scorer`? and not change the spacing?

ashutosh1919 replied:
Yep, we can do that but it will give pep8 error. @ogrisel suggested me to do like this [here](https://github.com/scikit-learn/scikit-learn/pull/15007#discussion_r382100880).

ogrisel replied:
Maybe it's fine to break PEP8 for this one. Using 2 spaces indent is not great either. Sorry about the back and forth.

You can put a `  # noqa` comment to silence the flake8 linter on that specific line.

ashutosh1919 replied:
Done

thomasjpfan reviewed (commented):
Thank you for the PR @ashutosh1919 !

I have a slight preference to having only `mean_absolute_relative_error` (not multiplying by 100).

ashutosh1919 committed:
Update sklearn/metrics/_regression.py

Co-Authored-By: Thomas J Fan <thomasjpfan@gmail.com>

ashutosh1919 committed:
Update sklearn/metrics/_regression.py

Co-Authored-By: Thomas J Fan <thomasjpfan@gmail.com>

ashutosh1919 committed:
Update sklearn/metrics/_regression.py

Co-Authored-By: Thomas J Fan <thomasjpfan@gmail.com>

ashutosh1919 committed:
Update sklearn/metrics/_regression.py

Co-Authored-By: Thomas J Fan <thomasjpfan@gmail.com>

ashutosh1919 commented:
> Thank you for the PR @ashutosh1919 !
> 
> I have a slight preference to having only `mean_absolute_relative_error` (not multiplying by 100).

@thomasjpfan , Thanks for the review!
Yes we know that we should have `mean_absolute_relative_error` in sklearn. But It's good to have both since both are used commonly. The benefit of using MAPE is that the user will be able to see error with more precision and with physical significance. @ogrisel has told me to work on that. After merge of this PR, I will work on that and I will submit PR for `mean_absolute_relative_error` too.

ogrisel commented:
> The benefit of using MAPE is that the user will be able to see error with more precision and with physical significance.

Well nothing else in scikit-learn is expressed in percentage: accuracy, balanced accuracy, precision, recall, f1 score, explained variance ... are all expressed in values between 0 and 1. So just for the sake of consistency I would also favor the raw `mean_absolute_relative_error` over the 100x scaled value.

And indeed I also prefer `mean_absolute_relative_error` (that's the one I would use to teach :) but I acknowledge that MAPE is apparently the referenced algorithm and it has "percentage" in its name unfortunately.

Also the formula in the [Wikipedia article](https://en.wikipedia.org/wiki/Mean_absolute_percentage_error) defines the MAPE with a formula that does not multiply by 100 but then notes:

>  Note that the MAPE is also sometimes reported as a percentage, which is the above equation multiplied by 100

So to summarize I would be:

- +1 on at least including  `mean_absolute_relative_error` without the 100x scale;
- +0 on also including `mean_absolute_percentage_error` that would just return  `mean_absolute_relative_error` multiplied by 100% for discoverability / convenience purposes.

But I would be also fine with just implementing  `mean_absolute_relative_error` if we mention both in the documentation, docstring and a few examples that use it that this is akin to MAPE / Mean Absolute Percentage Error divided by a factor of 100.

ogrisel commented on doc/modules/model_evaluation.rst:
```diff
@@ -1958,6 +1958,42 @@ function::
+
+.. math::
+
+  \text{MAPE}(y, \hat{y}) = \frac{\mid y - \hat{y} \mid}{\text{MAX}(\epsilon, \mid y \mid)}
```
Please be consistent with the formulas of Mean Absolute Error and Mean Squared Error upper in this user guide and include the sum over the per-sample errors:

```
\text{MAPE}(y, \hat{y}) = \frac{1}{n_{\text{samples}}} \sum_{i=0}^{n_{\text{samples}}-1} \frac{{}\left| y_i - \hat{y}_i \right|}{max(\epsilon, \left| y_i \right|)}
```

ashutosh1919 replied:
Done

ogrisel commented on doc/whats_new/v0.23.rst:
```diff
@@ -259,6 +259,12 @@ Changelog
+  the associated scorer for regression problems. :issue:`10708` fixed with the
+  PR :pr:`15007` by :user:`Ashutosh Hathidara <ashutosh1919>`. The scorer and
+  some practical test cases were taken from PR :pr:`10711` by
+  :user:`Mohamed Ali Jamaoui <mohamed-ali>`.
```
Let's not be finicky about credits :)

```
- |Feature| Added :func:`metrics.mean_absolute_percentage_error` metric and
  the associated scorer for regression problems. :issue:`10708` fixed with the
  PR :pr:`15007` by :user:`Ashutosh Hathidara <ashutosh1919>` and
  :user:`Mohamed Ali Jamaoui <mohamed-ali>`.
```

ogrisel reviewed (approved):
Apart from the above discussion about the naming / 100x scaling, this LGTM.

ashutosh1919 committed:
Modified Files to optimize changes

ashutosh1919 committed:
Merge remote-tracking branch 'upstream/master'

ashutosh1919 committed:
Modified Files to optimize changes

mohamed-ali commented on doc/whats_new/v0.23.rst:
```diff
@@ -269,8 +269,7 @@ Changelog
   the associated scorer for regression problems. :issue:`10708` fixed with the
-  PR :pr:`15007` by :user:`Ashutosh Hathidara <ashutosh1919>`. The scorer and
-  some practical test cases were taken from PR :pr:`10711` by
+  PR :pr:`15007` by :user:`Ashutosh Hathidara <ashutosh1919>` and
```
It would be useful to refer to the other PR as well (https://github.com/scikit-learn/scikit-learn/pull/10711)

ashutosh1919 replied:
@mohamed-ali , I have mentioned it previously, but @ogrisel suggested me to change this [here](https://github.com/scikit-learn/scikit-learn/pull/15007#discussion_r384657350).

ogrisel replied:
Ok sorry. The initial credit line by @ashutosh1919 was better then.

ashutosh1919 replied:
It's fine @ogrisel. Changed again. cc @mohamed-ali, Please look at the changes. I also need your approval since you were the first one working on this one.

ashutosh1919 committed:
Changed description of contributors info

ogrisel commented:
Before considering merging this PR I would love to make sure that @jnothman @lesteve and @amueller who participated in the review of #10711 agree with the current course of action, in particular with respect to the discussion about the naming and 100x scaling: https://github.com/scikit-learn/scikit-learn/pull/15007#issuecomment-591514981

ashutosh1919 commented:
@ogrisel , Please help me resolving this new error of `Pull Request Labeler / triage (pull_request)` failure. I have not seen this before on any commit.

ashutosh1919 commented:
> Before considering merging this PR I would love to make sure that @jnothman @lesteve and @amueller who participated in the review of #10711 agree with the current course of action, in particular with respect to the discussion about the naming and 100x scaling: [#15007 (comment)](https://github.com/scikit-learn/scikit-learn/pull/15007#issuecomment-591514981)

That will be great then. I also prefer 'mean_absolute_relative_error' over MAPE. So, if everyone here agrees to have only 'mean_absolute_relative_error' and not MAPE, then I will change files on this PR itself to create 'mean_absolute_relative_error'.

rth commented:
> Please help me resolving this new error of Pull Request Labeler / triage (pull_request) failure. I have not seen this before on any commit.

Ignore it for now: https://github.com/scikit-learn/scikit-learn/pull/16520

+1 to have a scaling between [0, 1].

ogrisel committed:
Merge branch 'master' into master

maikia committed:
DOC improve diabetes dataset description (#16534)

agramfort committed:
TST add test of fit attributes (#16286)

rth committed:
ENH Minimal Generalized linear models implementation (L2 + lbfgs) (#14300)

Co-authored-by: Christian Lorentzen <lorentzen.ch@googlemail.com>
Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>
Co-authored-by: Nicolas Hug <contact@nicolas-hug.com>

lschwetlick committed:
FIX Adress decomposition.PCA mle option problem (#16224)

jeremiedbb committed:
DOC add 0.22.2 in website news (#16631)

rth committed:
TST Enable california_housing pandas test in cron job (#16547)

rth committed:
EXA align lorenz curves between the two examples with GLMs (#16640)

adrinjalali committed:
DOC update n_jobs description in DBSCAN (#16615)

Co-authored-by: JohanWork

Commit:
FIX Pass sample_weight when predicting on stacked folds (#16539)

thomasjpfan committed:
BLD Turns off memory_profiler in examples to fix CircleCI (#16629)

thomasjpfan committed:
BLD Updates osx vm image in azure pipelines (#16647)

maurapintor committed:
FIX: normalizer l_inf should take maximum of absolute values (#16633)

cmarmo committed:
ENH Add check for non binary variables in OneHotEncoder. (#16585)

Co-authored-by: Thomas J Fan <thomasjpfan@gmail.com>
Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

merrcury committed:
DOC Update LICENSE Year (#16660)

lrjball committed:
BUG Fix issue with KernelPCA.inverse_transform (#16655)

cmarmo committed:
[MRG] DOC Add example about interpretation of coefficients of linear models (#15706)



Co-authored-by: Nicolas Hug <contact@nicolas-hug.com>
Co-authored-by: Gael Varoquaux <gael.varoquaux@normalesup.org>
Co-authored-by: Joel Nothman <joel.nothman@gmail.com>
Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

alexhenrie committed:
MNT Remove unused imports (#16665)

thomasjpfan committed:
MNT Restores behavior of conditioning on linting for most instances (#16672)

thomasjpfan committed:
BUG Fixes HistGradientBoosting when warm_start is on + early_stopping is on + no validation (#16663)

glemaitre committed:
BUG fix the math issue in latex compilation (#16673)

glemaitre committed:
BUG remove $ math env due to latex error (#16674)

nilichen committed:
DOC add example to tree.ExtraTreeClassifier (#16671)

Co-authored-by: Jonathan Rahn <jonathan.rahn@42digital.de>
Co-authored-by: Katrina Ni <kani@tableau.com>

ogrisel committed:
PEP8 in test_encoders.py

thomasjpfan committed:
MNT Removes unused private attributes (#16675)

rth committed:
CI Check for unused imports when linting (#16678)

* Check for missing imports in linting

* Exclude externals

* Remove unused imports

GaelVaroquaux committed:
DOC wording in linear model interpretation example (#16680)

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>
Co-authored-by: Chiara Marmo <cmarmo@users.noreply.github.com>
Co-authored-by: Nicolas Hug <contact@nicolas-hug.com>

adrinjalali committed:
API make __init__ params in cross_decomposition kw-only (#16682)

marenwestermann committed:
DOC Adds example to OAS (#16681)

ogrisel committed:
DOC Add note on bias induced by dropping categories in OneHotE… (#16679)

adrinjalali committed:
API make __init__ params in compose module kw-only (#16542)

* API make __init__ params in compose module kw-only

* pep8

* move * to after

ashutosh1919 committed:
Merge remote-tracking branch 'upstream/master'

ashutosh1919 commented:
Added `mean_absolute_relative_error` by PR #16689 .

agramfort commented on sklearn/metrics/_regression.py:
```diff
@@ -188,6 +190,78 @@ def mean_absolute_error(y_true, y_pred,
+    epsilon = np.finfo(np.float64).eps
+    mape = np.abs(y_pred - y_true) / np.maximum(np.abs(y_true), epsilon)
+    output_errors = np.average(mape,
+                               weights=sample_weight, axis=0) * 100.0
```
sticking to wikipedia I would remove the 100. See https://en.wikipedia.org/wiki/Mean_absolute_percentage_error

this also means that the other PR about mean absolute relative error is redundant and unnecessary.

ashutosh1919 replied:
In that case, I would suggest we should have one of the two: MAPE or MARE. My preference is MARE. I have created two PRs so that whatever senior contributors suggest, we can merge that one.

jnothman replied:
Let's remove the 100, and make sure this design decision is clearly noted.

ashutosh1919 replied:
Done

agramfort commented:
yes we need only one metric. I would go with MAPE to match wikipedia. MARE
would be our invention. We can
now make sure that google finds MAPE in sklearn doc when looking for mean
absolute relative error.

>

ogrisel commented:
@agramfort to clarify your point: you want to only implement the `mean_absolute_percentage_error` score function (and the scorer) but without the multiplication by 100 (and therefore not implement `mean_absolute_relative_error`).

We would need to make it extra explicit in the documentation (user guide and docstring) that our implementation does not make the multiplication by 100.

I think @rth also agrees with this.

My original position expressed in https://github.com/scikit-learn/scikit-learn/pull/15007#issuecomment-591514981 was a bit different, but if others (in particular @jnothman @lesteve and @amueller who participated in the review of #10711) also agree with this choice, then I am fine with it.

jnothman commented:
We could invent mean_absolute_proprotional_error to have the same acronym??!

agramfort commented:
> We could invent mean_absolute_proprotional_error to have the same acronym??!

I prefer not to me more creative than wikipedia :)
+1 for mean_absolute_percentage_error without the 100.

jnothman commented on doc/whats_new/_contributors.rst:
```diff
@@ -176,4 +176,6 @@
 .. _Nicolas Hug: https://github.com/NicolasHug
 
-.. _Guillaume Lemaitre: https://github.com/glemaitre
+.. _Ashutosh Hathidara: https://github.com/ashutosh1919
```
this is nowadays reserved for core devs whose names are entered frequently into the changelog. Usually we just use `:user:`. Please revert this.

ashutosh1919 replied:
Done

jnothman commented on sklearn/metrics/_regression.py:
```diff
@@ -188,6 +190,78 @@ def mean_absolute_error(y_true, y_pred,
+                                   sample_weight=None,
+                                   multioutput='uniform_average'):
+    """Mean absolute percentage error regression loss
+
```
Note here that we do not represent the output as a percentage: 1, rather than 100, indicates that the difference between y_pred and y_true is equal to y_true.

ashutosh1919 replied:
Done

jnothman commented on sklearn/metrics/_regression.py:
```diff
@@ -188,6 +190,78 @@ def mean_absolute_error(y_true, y_pred,
+
+        MAPE output is non-negative floating point. The best value is 0.0.
+        But note the fact that bad predictions can lead to arbitarily large
+        MAPE values, especially if some y_true values are very close to zero.
```
Note that we return a large value instead of `inf` when y_true is 0.

ashutosh1919 replied:
Done

jnothman reviewed (commented):
Otherwise this is looking good

ogrisel commented:
>  We could invent mean_absolute_proprotional_error to have the same acronym??!

It looks to much like a hack and feels un-natural to me. I think would rather either:
- stick to `mean_absolute_relative_error` and just mention the MAPE acronym and its expansion in the docstring and user guide instead.
- or just use the name `mean_absolute_percentage_error` without the x100 operation as Wikipedia does it.

agramfort commented:
>
>
>    - or just use the name mean_absolute_percentage_error without the x100
>    operation as Wikipedia does it.
>
>
+1

ashutosh1919 committed:
Removed x100 from MAPE and modified tests too

ashutosh1919 committed:
Merge remote-tracking branch 'upstream/master'

ashutosh1919 commented:
@jnothman , I have updated the code as per your last review. 
@agramfort, @ogrisel, @jnothman, @amueller, @rth - Note that I have removed x100 from the MAPE code and now it will give error for all valid inputs in the range [0,1]. Please review the latest code.

Now, the code in this PR and #16689 for MARE are same other than the difference of names.
If we are going ahead with this one. Then please close #16689 and the corresponding issue.

jnothman commented on sklearn/metrics/_regression.py:
```diff
@@ -188,6 +190,81 @@ def mean_absolute_error(y_true, y_pred,
+    """Mean absolute percentage error regression loss
+
+    Note here that we do not represent the output as a percentage in range
+    [0, 100]. Instead, we represent it in range [0, 1]. Read more in the
```
Surely the value is bounded from above by 1/eps, not 1?

ashutosh1919 replied:
done

ashutosh1919 committed:
Merge remote-tracking branch 'upstream/master'

ashutosh1919 committed:
Changed range

jnothman reviewed (commented):
One last think: `doc/modules/model_evaluation.rst` lists scorers. `neg_mean_absolute_percentage_error` is absent there.

rth committed:
Merge remote-tracking branch 'upstream/master' into ashutosh1919-master

rth committed:
DOC Add  absolute_percentage_error to doc/modules/model_evaluation.rst

rth commented:
I have addressed the last comment. Merging with +3. Thank you!

ShivKJ commented:
Wiki [mentions](https://en.wikipedia.org/wiki/Mean_absolute_percentage_error#:~:text=For%20forecasts%20which%20are%20too,limit%20to%20the%20percentage%20error.) multiplication by 100. Not Multiplying by 100 (current impl) seems so unnatural.

> > We could invent mean_absolute_proprotional_error to have the same acronym??!
> 
> It looks to much like a hack and feels un-natural to me. I think would rather either:
> 
> * stick to `mean_absolute_relative_error` and just mention the MAPE acronym and its expansion in the docstring and user guide instead.
> * or just use the name `mean_absolute_percentage_error` without the x100 operation as Wikipedia does it.
